### DISCUSSION TITLE
After blinking the LEDs. I started with the chapter register and on the first page itself I got some issues. I read about most of them and find out that the address is const and to use that fixed value we used raw pointer `GPIOE_BSRR as *mut u32` which is making a copy of the address(referencing) variable and allowing us to change.

### DISCUSSION THREAD

**Discussion by NitinSaxenait** [japaric/discovery#353] [Q&A] [answered]
After blinking the LEDs. I started with the chapter register and on the first page itself I got some issues. I read about most of them and find out that the address is const and to use that fixed value we used raw pointer `GPIOE_BSRR as *mut u32` which is making a copy of the address(referencing) variable and allowing us to change.
Second, we are dereferencing using * and putting the statement inside the unsafe block as per the raw pointer deference rules..

I hope till here i am getting it right now the next issue is what we are doing at right-hand side using the shift operator. We are using 9 11 to set and 16, 25 to reset **but why 1 in every shift?**

`#![no_main]
#![no_std]

#[allow(unused_imports)]
use aux7::{entry, iprint, iprintln};

#[entry]
 fn main() -> ! {
    aux7::init();

   {
        // A magic address!
        const GPIOE_BSRR: u32 = 0x48001018;

        // Turn on the "North" LED (red)
        //we use the dereference operator * on a raw pointer that requires an unsafe block.
        //change data through the mutable pointer
        (GPIOE_BSRR as *mut u32) = 1 << 9;

        // Turn on the "East" LED (green)
        *(GPIOE_BSRR as *mut u32) = 1 << 11;

        // Turn off the "North" LED
        (GPIOE_BSRR as *mut u32) = 1 << (9 + 16);

        // Turn off the "East" LED
        (GPIOE_BSRR as *mut u32) = 1 << (11 + 16);
    }
    loop {}
}
`
@adamgreig

_Originally posted by @NitinSaxenait in https://github.com/rust-embedded/discovery/issues/193#issuecomment-864730022_

**Comment by adamgreig** [comment, marked as answer]
> but why 1 in every shift?

It's a convenient way to express the full binary number we want to write to the register. The way the BSRR register works is you write a '1' to any of the 'set' fields (each 1 bit wide, bits 0-15) to turn on that GPIO, and write a '1' to any of the 'reset' fields (bits 16-31) to turn it off.

So, to turn on pin 2, you write a 1 to bit 2, which means writing `0b0000_0000_0000_0000_0000_0000_0000_0100` to the register. Instead of typing out the long number, we can write `1 << 2`, which is the same number but easier to understand. We're shifting a 1 to the left by 2 places. Because all the fields can only take a '1', we never need any other number. To turn it off we'd need to write `0b0000_0000_0000_0100_0000_0000_0000_0000`, or for short `1 << 18`, but we write it `1 << (2 + 16)` to make it clearer that it's the second bit, just shifted by an extra 16 positions to turn it off.

Sometimes a field is longer, for example it might be 4 bits starting at bit 5, and we want to write `0b1010` to that field: overall we'd need to write `0b0000_0000_0000_0000_0000_0001_0100_0000`, but instead we could say `0b1010 << 5` or `6 << 5` which is the same thing.

Hope that helps answer your question!

**Reply by NitinSaxenait** [reply]
Yes thanks, @adamgreig for this. It is clear now.

**Reply by NitinSaxenait** [reply]
@adamgreig I have one more doubt. Do we still need to use the aux7 crate as if I remove it from the main program, the program doesn't work correctly. Can I consider LEDs as a room and to access them I need to go through the main door(device_peripheral)??
